### PR TITLE
Collect GDPR and CCPA Consent State

### PR DESCRIPTION
**Pull request by alexs-mparticle**

### CODE REVIEW

File: mParticleCore.brs [hunk 1]
```diff
@@ -138,6 +138,94 @@ function mParticleConstants() as object
...
+        end function,
+
+        ' GDPR
+        addGDPRConsentState : function (consentState as object, purpose as string, gdprConsent as object)
+            consentState.gdpr.AddReplace(purpose, gdprConsent)
+        end function,
+        getCCPAConsentState : function (consentState as object)
+            return consentState.ccpa.Lookup("data_sale_opt_out")
+        end function,
+        getGDPRConsentState : function (consentState as object, purpose as string)
```
Comment (rmi22186) [inline mParticleCore.brs]: `getGDPRConsentState` shouldn't take any arguments. it should just return the GDPRConsentState.  it's up to the developer if they want to then lookup a purpose on it

File: mParticleCore.brs [hunk 2]
```diff
@@ -1341,6 +1452,10 @@ function mParticleStart(options as object, messagePort as object)
         logEvent:           function(eventName as string, eventType = mParticleConstants().CUSTOM_EVENT_TYPE.OTHER, customAttributes = {}) as void
                                 m.logMessage(m.model.CustomEvent(eventName, eventType, customAttributes))
                             end function,
+        sendUserAttributeChangeEvent: function(attributeKey as string, newUserAttributeValue as string, previousUserAttribute as string, isNewAttribute as boolean, deleted as boolean) as void
```
Comment (rmi22186) [inline mParticleCore.brs]: i think this is part of the leanplum bug, you can probably remove

File: mParticleCore.brs [hunk 3]
```diff
@@ -448,6 +540,20 @@ function mParticleStart(options as object, messagePort as object)
             return userAttributes
         end function
         
+        storage.setConsentState = function(mpid as string, consentState as object) as void
```
Comment (rmi22186) [inline mParticleCore.brs:537]: this block of code is a bit confusing to me. let's review in person

**Review by rmi22186 — COMMENTED** [review]
looks good. small nits. but i'd like to grab some time to review a couple small lines in person which i noted, and also once consent is granted to the specific workspace, would be great to see it flowing end to end.

File: mParticleCore.brs [hunk 4]
```diff
@@ -138,6 +138,94 @@ function mParticleConstants() as object
...
+        end function,
+        setTimestamp : function (ccpaConsentState as object, timestamp as LongInteger)
+            ccpaConsentState.ts = timestamp
+        end function,
+        setHardwareId : function (ccpaConsentState as object, hardwareId as string)
+            ccpaConsentState.h = hardwareId
+        end function,
+    }
+    GDPRConsentState = {
+        build : function (document="" as string, consented=false as boolean, location="" as string, timestamp=0 as longInteger, hardwareId="" as string)
```
Comment (samdozor) [inline mParticleCore.brs]: is there any reason to default these to empty string? better to upload them as null if not included. same comment above on line 177.

File: mParticleCore.brs [hunk 5]
```diff
@@ -448,6 +540,20 @@ function mParticleStart(options as object, messagePort as object)
             return userAttributes
         end function
         
+        storage.setConsentState = function(mpid as string, consentState as object) as void
+            m.set(m.mpkeys.CONSENT_STATE + mpid, FormatJson(consentState))
+            m.flush()
+        end function
+
+        storage.getConsentState = function(mpid as string) as object
+            attributeJson = m.get(m.mpkeys.CONSENT_STATE + mpid)
```
Comment (samdozor) [inline mParticleCore.brs]:
```suggestion
            consentJson = m.get(m.mpkeys.CONSENT_STATE + mpid)
```

**Review by samdozor — CHANGES_REQUESTED** [review]

**Comment by rmi22186** [comment]
@samdozor - any final feedback on this?

**Review by samdozor — APPROVED** [review]

**Pull request merged**